src/app/OperationalDeviceProxy.cpp

```diff
@@ -212,40 +203,40 @@ void OperationalDeviceProxy::DequeueConnectionFailureCallbacks(CHIP_ERROR error,
     }
 }
 
-void OperationalDeviceProxy::OnSessionEstablishmentError(CHIP_ERROR error)
+void OperationalDeviceProxy::HandleCASEConnectionFailure(void * context, CASEClient * client, CHIP_ERROR error)
 {
-    VerifyOrReturn(mState != State::Uninitialized && mState != State::NeedsAddress,
-                   ChipLogError(Controller, "OnSessionEstablishmentError was called while the device was not initialized"));
+    OperationalDeviceProxy * device = static_cast<OperationalDeviceProxy *>(context);
+    VerifyOrReturn(device->mState != State::Uninitialized && device->mState != State::NeedsAddress,
+                   ChipLogError(Controller, "HandleCASEConnectionFailure was called while the device was not initialized"));
+    VerifyOrReturn(client == device->mCASEClient, ChipLogError(Controller, "HandleCASEConnectionFailure for unknown CASEClient"));
 
-    mState = State::Initialized;
-    mInitParams.idAllocator->Free(mCASESession.GetLocalSessionId());
+    device->mState = State::Initialized;
 
-    DequeueConnectionSuccessCallbacks(/* executeCallback */ false);
-    DequeueConnectionFailureCallbacks(error, /* executeCallback */ true);
+    device->DequeueConnectionSuccessCallbacks(/* executeCallback */ false);
+    device->DequeueConnectionFailureCallbacks(error, /* executeCallback */ true);
+    device->DeferCloseCASESession();
 }
 
-void OperationalDeviceProxy::OnSessionEstablished()
+void OperationalDeviceProxy::HandleCASEConnected(void * context, CASEClient * client)
 {
-    VerifyOrReturn(mState != State::Uninitialized,
-                   ChipLogError(Controller, "OnSessionEstablished was called while the device was not initialized"));
+    OperationalDeviceProxy * device = static_cast<OperationalDeviceProxy *>(context);
+    VerifyOrReturn(device->mState != State::Uninitialized,
+                   ChipLogError(Controller, "HandleCASEConnected was called while the device was not initialized"));
+    VerifyOrReturn(client == device->mCASEClient, ChipLogError(Controller, "HandleCASEConnected for unknown CASEClient"));
 
-    // TODO Update the MRP params based on the MRP params extracted from CASE, when this is available.
-    CHIP_ERROR err = mInitParams.sessionManager->NewPairing(
-        Optional<Transport::PeerAddress>::Value(mDeviceAddress), mPeerId.GetNodeId(), &mCASESession,
-        CryptoContext::SessionRole::kInitiator, mInitParams.fabricInfo->GetFabricIndex());
+    CHIP_ERROR err = client->DeriveSecureSessionHandle(device->mSecureSession);
     if (err != CHIP_NO_ERROR)
     {
-        ChipLogError(Controller, "Failed in setting up CASE secure channel: err %s", ErrorStr(err));
-        OnSessionEstablishmentError(err);
-        return;
+        device->OnSessionEstablishmentError(err);
```

Comment:
I think mCASEClient may be leaked here.  Which OnSessionEstablishmentError is this line calling?

Reply:
Apparently fixed with replacement of this call with HandleCASEConnectionFailure.

Thanks.